doc/sysconf.7
.Au John Kaul
.Dd May 10 2026
.Dt sysconf.7
.Os
.Pp
.Sh NAME 
.Nm sysconf 
.Nd A dependency free key/value configuration file editing tool.
.Pp
.Sh SYNOPSIS 
.Nm
-f file.conf
.Nm
-f file.conf [-d file.conf.defaults]
.Nm
-f file.conf [-n] [key]
.Nm
-f file.conf [key=value]
.Nm
-f file.conf [key+=value]
.Nm
-f file.conf [key-=value]
.Pp
.Sh OPTIONS 
.Bl -tag -width Ds
.It Fl f Ar file.conf
Configuration file
.Pp
.It Fl d Ar file.conf.defaults
Check the 
.Li file.conf
's key values against 
.Li file.conf.defaults
's key
values for duplicate entries.
.Pp
.It Fl n
Display "key" as well when retrieving a variable. The default method
is to only display a key's value but this option makes the return show
both the key and the value.
.Pp
.It key
Displays the values associated with given key.
.Pp
.It key=value
This option adds a new or overwrites an existing key/value entry.
.Pp
.It key+=value
This option add to an existing key/value entry.
.Pp
.It key-=value
This option deletes an existing key's value entry. If the last value
is called to be removed the entire key/value is removed from the file.
.El
.Pp
.Sh DESCRIPTION 
This utility will 'get' (-i.e., print to 
.Li stdout
) and/or 'set' (i.e.,
change) a key/value value stored in a configuration file formatted in
the typical key/value syntax with zero dependencies on external
libraries or tools. There are several popular key/value styles
different programs and tools can use so this utility was designed to
tokenize several.
.Pp
By design 
.Nm
does not have any flags to specify actions to preform
different actions, instead this utility will try to determine what to
do, or not to do, based on the arguments given and what is currently
in the configuration file. See the examples below. During a 'set'
operation, 
.Nm
will also try to determine the syntax of the key/value
pair, to be changed, by reading the key/value string (-e.g. based on
the key/value being changed, 
.Nm
will try to determine if the value
needs quotes or a termination character and either spaces or an equal
sign or colon separation character).
.Pp
This utility can also add and delete values to a key with 
.Cm +=
and 
.Cm -=.
.Pp
Example configuration file syntaxes this utility can 'get' and 'set'.
.Bd -literal -offset indent
    key=value
    key = value;
    key.subkey = value;
    key="value"
    key value
    key value;
    key:"value"
    key = "value1 value2";
    etc.
.Ed
.Pp
.Sy NOTE
:
Some keys or values in configuration files can contain special
characters like dollar signs ($). To pass these characters to this
utility, the dollar sign must be escaped with a slash (\\) or the
string can be surrounded with single quotes (').
.Pp
If the configuration file does not already contain the key/value entry
when a call to 'set' is made, this utility will add the new entry, at
the bottom of the file, in the following format (which is also the
default syntax for rc.conf key/values). If the configuration file does
not already exist, 
.Nm
will create the file and add the key/value
entry if called with the 
.Cm =
operator.
.Pp
.Bd -literal -offset indent
    % sysconf -f /path/missing.conf key=value
.Ed
.Pp
This utility does not allow/account for duplicate key/value entries in
the 'get' and 'set' stages. The files first key/value entry found
during a 'get' is returned and the files first key/value entry is
written to the 'set' stage (the second--duplicate--key/value is not
written).
.Pp
It is also possible to check a configuration file's key/values against
a default configuration file which will search for duplicate values in
keys listed in the configuration file. For example, in FreeBSD the
/etc/rc.conf file is included from the file /etc/defaults/rc.conf,
which specifies the default settings for all the available options.
Options need only be specified in /etc/rc.conf when the system
administrator wishes to override the defaults in
/etc/defaults/rc.conf. See the example section below for 'checking for
duplicates'.
.Pp
.Pp
.Sh EXAMPLES 
.Pp
.Em LIST ALL VALUE(S)
.Pp
To list off (pretty print) the key/values in a configuration file just
call the program with no arguments. However, the way this program
parses the key/value strings may result in slightly odd looking
results so you should not try to use this utility to remove duplicate
key/values.
.Pp
.Bd -literal -offset indent
    % syconf -f /path/file.conf
.Ed
.Pp
.Em GET A VALUE
.Pp
To retrieve the value from a key in a file call this program with just
the key to search for. This utility will only return the value(s) for
the given key to better support usage from within a script.
.Bd -literal -offset indent
    % sysconf -f /path/file.conf key
    % value
.Ed
.Pp
If there is a need to produce the key as well as the value in the
output of a 'get a value' operation, the 
.Li -n
switch can be used.
.Bd -literal -offset indent
    % sysconf -f /path/file.conf -n key
    % key: value
.Ed
.Pp
.Em CHANGE VALUE(S)
.Pp
To change a value associated with a key use the equal ( 
.Cm =
) sign. A
change will only be made if the value does not match what is currently
in the configuration file -i.e., if the value to be assigned is already the
same as it is in the configuration file, no change is made.
.Bd -literal -offset indent
    % sysconf -f /path/file.conf key=value
.Ed
.Pp
To add to a value associated with a key use a plus-equal ( 
.Cm +=
). A
change will only be added if the value does not already exist in the
configuration value string -i.e., if the value to be assigned is already part
of the configuration file value, no change is made.
.Bd -literal -offset indent
    % sysconf -f /path/file.conf key+=value
.Ed
.Pp
To remove a value associated with a key use a minus-equal ( 
.Cm -=
). A
change will only be made if the value exists in the configuration file value
string. And when the last value for the given key is called out to be
removed the key will also be removed from the file.
.Bd -literal -offset indent
    % sysconf -f /path/file.conf key-=value
.Ed
.Pp
.Em ESCAPING CHARS
.Pp
To use a dollar sign in a key, escape it.
.Bd -literal -offset indent
    % sysconf -f /path/file.conf \\$key
.Ed
.Pp
or use single quotes to surround the string.
.Bd -literal -offset indent
    % sysconf -f /path/file.conf '$key'
.Ed
.Pp
.Em CHECKING FOR DUPLICATES
.Pp
To search for duplicate values in a configuration file against a
default configuration file key values, use the -f and -d flags.
.Bd -literal -offset indent
    % sysconf -f /path/file.conf -d /path/file.conf.defaults
.Ed
.Pp
.Sh EXIT 
The 
.Nm
utility exits 0 on success, and >0 if an error occurs.
.Pp
.Sh HISTORY 
Created for my personal use.
.Pp
.Sh AUTHOR 
John Kaul


changelog.md
# Changelog

- process CLI args with getopt #12

v1.3.2 - 2026-05-10
- Fixed program return codes. 
  program now exits 0 on success, and >0 if an error occurs.
- Added more unit tests (credit: @msimerson).

v1.3.1 - 2026-05-07
- Added macros to call for cleanup and errors. This cleans up the code
  a bit and makes it more standarized (less code duplication and
  errors, now consitatly print to stderr).
- Added support to supply arguments in any order. This allows the
  ability to call the program in a number of different ways such as:
```
     % sysconf -n -f <file.conf> key=value
     % sysconf -f <file.conf> -n key=value
     % sysconf key=value -n -f <file.conf>
     ...
```
- Changed argument count index to 1 (instead of 0) to skip program
  name (index 0).
- Added a github workflow test.yml file from: @msimerson to build and
  run tests.

v1.3.0 - 2026-05-05
- Added ability to handle indented key/value strings.

v1.2.0 - 2026-05-04
- SoC: Moved printing (to STDOUT) from `sysconf.c` to the `print-config.c`.
- Cleaned up memory managment.
- Removed reference to "Semantic Versioning". This should allow me to
  version any code related changes that do not add features or are for
  bug/issue fixes.

v1.1.2 - 2026-05-04
- Corrected config file creating portion whereas file was immediately
  closed.
- Corrected issue where strings were not being matched exactly in
  either key or value.

v1.1.1 - 2026-04-29
- Corrected read-after-free bug.

v1.1.0 - 2026-04-28
- Updated documentation.

v1.0.0 - 2026-04-28
- Introduced static versioning via src/version.h
- Added CHANGELOG.md for project tracking
- Removed Git-based version generation from the Makefile


src/sysconf.c
//===---------------------------------------------------*- C -*---===
// File Last Updated: 05.04.26 21:35:15
//
//: main.c
//
// BY  : John Kaul [john.kaul@outlook.com]
//
// DESCRIPTION
// This is a utility to list and change key/value based config file values.
//
// Configuration files typically have the `key = value` syntax but
// some key/value items are seperated with equal (=), spaces ( ),
// colons (:), etc. and there does't seem to be a utility that allows
// for quick changing of these values.
//
// The `sysrc` utility in FreeBSD allows for easy changing of
// variables/values but the utility does not work with doted names
// -i.e. `key.subkey = value;` entries. This utility should help.
//
// This utility can also be called from a script; this utility when
// reporting the value for a key only reports the value for the given
// variable (like using the -n switch with sysrc).
//
// This utility can preform actions based on the arguments given.
// For example:
//      % sysconf -f <config_file>
//    Will display the cofig_file key=value pair.
//
//      % sysconf -f <config_file> key
//    Will display the config_file key's value.
//
//      % sysconf -f <config_file> key=value
//      % sysconf -f <config_file> key+=value
//      % sysconf -f <config_file> key-=value
//    Will change the config_file value to the value specified as an argument.
//
//      % sysconf -f <config_file> -d <defaults_config_file>
//    Will check for duplicate value entries for each key in the config_file against the
//    defaults_config_file.
//
// If this utlity is called to set a key/value and the configuration
// file doesn't exist, it will be created.
//
// SYNOPSYS
//      sysconf -f configfile
//      sysconf -f configfile -d configfile.defaults
//      sysconf -f configfile [-n] [key]
//      sysconf -f configfile [key=value]
//      sysconf -f configfile [key+=value]
//      sysconf -f configfile [key-=value]
//===-------------------------------------------------------------===

#include "parse-config.h"
#include "print-config.h"
#include "version.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>

//---[ MACROS ]------------------------------------------------------
#define clean_argarray()                                        \
  do {                                                          \
    if(arg_array) {                                             \
      if(arg_count) {                                           \
        for (int i = 0; i < arg_count; i++) free(arg_array[i]); \
      }                                                         \
      free(arg_array);                                          \
      arg_array = NULL;                                         \
    }                                                           \
  } while (0)

#define clean_configarray()                                     \
  do {                                                          \
    if(config_array) {                                          \
      free_config(config_array, config_count);                  \
      free(config_array);                                       \
      config_array = NULL;                                      \
    }                                                           \
  } while (0)

#define cleanup()                                               \
  do {                                                          \
    clean_argarray();                                           \
    clean_configarray();                                        \
  } while (0)

#define err(errmsg)                                             \
  do {                                                          \
    fprintf(stderr, "%s *ERROR*: %s", argv[0], errmsg);         \
    cleanup();                                                  \
  } while (0)

#define usage()                                                 \
  do {                                                          \
    fprintf(stderr, "Version: %s\n", program_version);          \
    fprintf(stderr, "Usage: %s -f file.conf [-d file.defaults] [-n] [key[=value]]\n", argv[0]); \
  } while (0)

//------------------------------------------------------*- C -*------
// Main
//
// ARGS
//  argc            :   number of args
//  argv[]          :   array of arguments
//
// RETURN
//  int
//-------------------------------------------------------------------
int main(int argc, char *argv[]) {

  char *file_string = NULL;                             /* Used to store config file name */
  char *default_string = NULL;                          /* Used to store default config file name. */
  char *arg_string = NULL;                              /* Used to store the argument string. */
  char delimiters[] = " \t\n\"\':=;";
  int keyvalue_output = 0;

  // -Check the command line arguments.
  //  if there are not enough arguments, exit.
  if (argc < 3) {
    usage();
    fprintf(stderr, "Error: Invalid command line arguments\n");
    return 1;
  }

  // -Parse the command line options.
  for (int i = 1; i < argc; i++) {
    if (argv[i] && strlen(argv[i]) > 1) {
      if (argv[i][0] != '-') { arg_string = argv[i]; }
      if (argv[i][0] == '-' && argv[i][1] == 'f') { file_string = argv[++i]; }
      if (argv[i][0] == '-' && argv[i][1] == 'd') { default_string = argv[++i]; }
      if (argv[i][0] == '-' && argv[i][1] == 'n') { keyvalue_output = 1; }
    }
  }

  // -If there is not a `file_string` variable, quit.
  if (! file_string) {
    usage();
    fprintf(stderr, "Error: No configuration file to edit specified\n");
    return 1;
  }

  // -Keep a record of how many items in the config file.
  int config_count = 0;
  int arg_count = 0;

  // -Parse the config file.
  config_t* config_array = parse_config(file_string, &config_count, delimiters);

  // -If we couldn't parse the file, quit.
  if (!config_array) {
    fprintf(stderr, "Failed to parse the configuration file.\n");
    free(config_array);
    return 1;
  }

  if (default_string != NULL) {         /* We are going to check the config file for
                                         * duplicates against a defaults config file.
                                         */
    int default_count = 0;
    config_t* default_array = parse_config(default_string, &default_count, delimiters);

    if (!default_array) {
      fprintf(stderr, "Failed to parse the configuration file.\n");
      free(config_array);
      free(default_array);
      return 1;
    }

    for (int i = 0; i < config_count; i++) {
      if (config_array[i].values != NULL && config_array[i].values[0] != NULL) {
        char* key = config_array[i].values[0];
        config_t* temp_conf = find_config_item(config_array, key, config_count);
        config_t* temp_default = find_config_item(default_array, key, default_count);
        if (temp_default != NULL) {
          for (int x = 1; x < config_array[i].value_count; x++) {
            char* value = temp_conf->values[x];
            if (memcmp(value, "#", 1) == 0)
              break;
            if (contains(temp_default->values, temp_default->value_count, value) == 1)
              printf("*DUPLICATE* %s: '%s'\n", key, value);
          }
        }
      }
    }
    return 0;
  }     /* end_ if(default_string) */

  // -No argument (key = value or key) given so just
  //  print the config values.
  if(argc == 3) {
    printconfigfile(config_array, config_count);
    free_config(config_array, config_count);
    free(config_array);
    return 0;
  }

  // -Parse the argument string passed to this program.
  //  Based on the size of this array, we are going to determine if we
  //  need to preform replacement operations or just list the value.
  char **arg_array;                                     /* Used to store the argument
                                                           string passed to this program. */
  arg_count = make_argv(arg_string, delimiters, &arg_array);

  // -Do things differently based on the number of arguments given.
  //  no argument;
  //    disply the config file.
  //  key;
  //    look up the key in the config_file and disply it's value.
  //  key=value;
  //    Make a addition/replacement/update in the config_file.
  if (arg_count >= 1) {

    // Get the values associated with the argument passed to this function.
    char **config_line_array = get_value(config_array, config_count, arg_array[0]);

    // If the key cannot be found in the config file, we need to check
    // to see if the value is:
    // 1. being searched for; if so, exit.
    // 2. wanting to be added; if so, add it.
    if (config_line_array == NULL) {

      if (arg_count == 1) {                             /* Seems to be a simple search situation,
                                                           since, we do not have a key in the config
                                                           file, we exit.*/
        err("Error: key not found\n");
        return 1;
      }
      if (arg_count > 1) {                               /* Seems to be a condition where the key/value
                                                           needs to be appended to the config file.*/
        /* In the condition where the key is not found we need to
         * check to see if the string is not a += or -= operation
         * before we append the config file.  */
        if ((strnstr(arg_array[0], "-", strlen(arg_array[0])) != NULL) ||
            (strnstr(arg_array[0], "+", strlen(arg_array[0])) != NULL)) {
          err("Incorrect syntax. Key is not found in config file.\n");
          return 1;
        }

        writevariable(arg_array[0], arg_array, arg_count, file_string);

        cleanup();
        return 0;
      }
    }

    // -If we only have a 'key' as an argument; just find in the
    //  config and display the value set.
    if(config_line_array != NULL && arg_count == 1) {
      // Itterate the array and print chars.
      if (keyvalue_output != 0) {
        printf("%s: ", config_line_array[0]);
      }
      config_line_array += 1;                           /* strip the 'key' from the array */
      while(*config_line_array) {
        if (memcmp(*config_line_array, "#", 1) == 0)
          break;
        printf("%s ", *config_line_array++);
      }
      printf("\n");

      cleanup();
      return 0;
    }

    // -If the argument is equal to 'key=value' make a replacement.
    // -However, if the argument is 'key+=value' or 'key-=value' make an update.
    if(arg_count > 1) {

      // Count the config_line_array number of elements.
      int i = 0;
      for (; config_line_array[i] != NULL; i++)
        ;

      // -Determine if we have a change to make; compare the `arg_array`
      //  and the `config_array` values.
      if (contains(config_line_array, i, arg_array[1]) == 0) {          /* 0 = A value was not found in the key's string... */

        if (strstr(arg_array[0], "-") != NULL) {                        /* However, if the user wants to subtract a value
                                                                           but the value was not found, we need to exit. */
          err("Value not found in value string. No change made.\n");
          return 0;
        }

        replacevariable(config_line_array[0], arg_array, arg_count, file_string);

        /* LOGIC:
         * - The concept is to free `arg_array` up to the `arg_count`, and allow the
         *   `print-config.c` to free the additional value(s) added to the `arg_array`.
         */
        arg_array = NULL;

        cleanup();
        return 0;

      } else if(contains(config_line_array, i, arg_array[1]) == 1) {    /* 1 = Value found... */

        if (strstr(arg_array[0], "+") != NULL) {                        /* However, if the user wants to add a value
                                                                           for a value that was found, we need to exit. */
          err("Value found in key's value string. No change made.\n");
          return 0;
        }

        replacevariable(config_line_array[0], arg_array, arg_count, file_string);

        cleanup();
        return 0;
      } else {                                                    /* Assume the user wants to set a value that already exits. */
        printf("Value found. No change made.\n");
      }
    }   /* end_ if(arg_count > 1)  */
  }     /* end_ if(arg_count >= 1) */

  cleanup();
  return 0;
} ///:~
